alica_engine/src/engine/teammanager/TeamManager.cpp

```diff
@@ -138,49 +192,161 @@ bool TeamManager::isAgentActive(AgentIDConstPtr agentId) const
-    if (agentEntry != _agents.end()) {
-        return agentEntry->second->isIgnored();
-    } else {
-        return false;
+    Agent* agent = getAgent(agentId);
+    if (agent) {
+        return agent->isIgnored();
     }
+    // Ignore agents that are not announced yet
+    return true;
 }
 
 void TeamManager::setAgentIgnored(AgentIDConstPtr agentId, const bool ignored) const
 {
-    auto agentEntry = _agents.find(agentId);
-    if (agentEntry != _agents.end()) {
-        agentEntry->second->setIgnored(ignored);
+    Agent* agent = getAgent(agentId);
+    if (agent) {
+        agent->setIgnored(ignored);
     }
 }
 
 bool TeamManager::setSuccess(AgentIDConstPtr agentId, const AbstractPlan* plan, const EntryPoint* entryPoint)
 {
-    auto agentEntry = _agents.find(agentId);
-    if (agentEntry != _agents.end()) {
-        agentEntry->second->setSuccess(plan, entryPoint);
+    Agent* agent = getAgent(agentId);
+    if (agent) {
+        agent->setSuccess(plan, entryPoint);
         return true;
     }
     return false;
 }
 
 bool TeamManager::setSuccessMarks(AgentIDConstPtr agentId, const IdGrp& suceededEps)
 {
-    auto agentEntry = _agents.find(agentId);
-    if (agentEntry != _agents.end()) {
-        agentEntry->second->setSuccessMarks(suceededEps);
+    Agent* agent = getAgent(agentId);
+    if (agent) {
+        agent->setSuccessMarks(suceededEps);
         return true;
     }
     return false;
 }
 
 const DomainVariable* TeamManager::getDomainVariable(AgentIDConstPtr agentId, const std::string& sort) const
 {
-    auto agentEntry = _agents.find(agentId);
-    if (agentEntry != _agents.end()) {
-        return agentEntry->second->getDomainVariable(sort);
+    Agent* agent = getAgent(agentId);
+    if (agent) {
+        return agent->getDomainVariable(sort);
     }
     return nullptr;
 }
 
+bool TeamManager::updateAgents(AgentGrp& deactivatedAgents)
+{
+    AgentsCache::AgentMap& agents = *_agentsCache.get();
+    bool changedSomeAgent = false;
+    for (const auto& agent : agents) {
+        bool changedCurrentAgent = agent.second->update();
+        if (changedCurrentAgent && !agent.second->isActive()) {
+            deactivatedAgents.push_back(agent.second->getId());
+        }
+        changedSomeAgent |= changedCurrentAgent;
+    }
+    return changedSomeAgent;
+}
+
+void TeamManager::handleAgentQuery(const AgentQuery& aq) const
+{
+    if (!_useAutoDiscovery || aq.senderID == _localAgent->getId()) {
+        return;
+    }
+
+    const Agent* ag = _engine->getTeamManager().getAgent(aq.senderID);
+    if (ag && ag->isIgnored()) {
+        // if agent is already discovered and ignored
+        return;
+    }
+
+    // TODO: Add sdk compatibility check with comparing major version numbers
+    if (aq.senderSdk != _localAgent->getSdk() || aq.planHash != _localAgent->getPlanHash()) {
+        ALICA_DEBUG_MSG("tm: Version mismatch ignoring: " << aq.senderID << " sdk: " << aq.senderSdk << " ph: " << aq.planHash);
```

Comment:
To make the user's life easier: "ph" -> "plan hash".
"ph" requires previous knowledge. "plan hash" is self-evident.